src/controllers/suggestions.js

```diff
@@ -0,0 +1,385 @@
+    }
+
+    if (!isArray(context.data)) {
+      return badRequest('Request body must be an array');
+    }
+
+    const suggestionPromises = context.data.map(async (suggData, index) => {
+      try {
+        // eslint-disable-next-line no-param-reassign
+        suggData.opportunityId = opptyId;
+        const suggestionEntity = await Suggestion.create(suggData);
+        return {
+          index,
+          suggestion: SuggestionDto.toJSON(suggestionEntity),
+          statusCode: 201,
+        };
+      } catch (error) {
+        return {
+          index,
+          message: error.message,
+          statusCode: error instanceof ValidationError ? 400 : 500,
+        };
+      }
+    });
+
+    const responses = await Promise.all(suggestionPromises);
+    // Sort the results by the index of the suggestion in the request
+    responses.sort((a, b) => a.index - b.index);
+    const succeded = responses.filter((r) => r.statusCode === 201).length;
+    const fullResponse = {
+      suggestions: responses,
+      metadata: {
+        total: responses.length,
+        success: succeded,
+        failed: responses.length - succeded,
+      },
+    };
+
+    return createResponse(fullResponse, 207);
+  };
+
+  /**
+   * Updates data for a suggestion
+   * @param {Object} context of the request
+   * @returns {Promise<Response>} the updated suggestion data
+   */
+  const patchSuggestion = async (context) => {
+    const siteId = context.params?.siteId;
+    const opportunityId = context.params?.opportunityId;
+    const suggestionId = context.params?.suggestionId;
+
+    if (!hasText(siteId)) {
+      return badRequest('Site ID required');
+    }
+
+    if (!hasText(opportunityId)) {
+      return badRequest('Opportunity ID required');
+    }
+
+    if (!hasText(suggestionId)) {
+      return badRequest('Suggestion ID required');
+    }
+
+    const suggestion = await Suggestion.findById(suggestionId);
+    if (!suggestion || suggestion.getOpportunityId() !== opportunityId) {
+      return notFound('Suggestion not found');
+    }
+    const opportunity = await suggestion.getOpportunity();
+    if (!opportunity || opportunity.getSiteId() !== siteId) {
+      return notFound('Suggestion not found');
+    }
+
+    // validate request body
+    if (!context.data) {
+      return badRequest('No updates provided');
+    }
+
+    let hasUpdates = false;
+    const { rank, data, kpiDeltas } = context.data;
+    try {
+      if (rank && rank !== suggestion.rank) {
+        hasUpdates = true;
+        suggestion.setRank(rank);
+      }
+
+      if (data) {
+        hasUpdates = true;
+        suggestion.setData(data);
+      }
+
+      if (kpiDeltas) {
+        hasUpdates = true;
+        suggestion.setKpiDeltas(kpiDeltas);
+      }
+
+      if (hasUpdates) {
+        const updatedSuggestion = await suggestion.save();
+        return ok(SuggestionDto.toJSON(updatedSuggestion));
+      }
+    } catch (e) {
+      return badRequest(e.message);
```

Comment:
bad request would only be applicable if a validation error occurs, but there might also be other errors